### PR TITLE
usecases: add vsock documentation

### LINKED ISSUE #226: Add vsock documentation
Add vsock documentation

### PR DESCRIPTION
**Pull request by devimc**
Add vsock documentation

fixes #226

Signed-off-by: Julio Montes <julio.montes@intel.com>

### CODE REVIEW

File: use-cases/using-vsocks.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    - [proxy communication diagram](#proxy-communication-diagram)
+    - [vsock communication diagram](#vsock-communication-diagram)
+- [System requirements](#system-requirements)
+- [Advantages of using vsocks](#advantages-of-using-vsocks)
+    - [High density](#high-density)
+    - [Reliability](#reliability)
+
+## Introduction
+
+There are two different ways to communicate processes in the virtual machine with
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: I'd reword `to communicate processes`, maybe reflow the sentance. Suggestion:

> There are two different ways processes in the virtual machine can communicate with processes in the host.

File: use-cases/using-vsocks.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+processes in the host. The first one is by using serial ports, the processes in
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: s/the process/where the process/

File: use-cases/using-vsocks.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+the virtual machine can read/write data from/to a serial port device and the
+processes in the host can read/write data from/to a unix socket. This solution
+has been part of QEMU from long time ago and almost all GNU/Linux distributions
+have support for serial ports, making it the most portable solution. However,
+there is a limitation that impact in its usage and implementation, the
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: s/impact in its/impacts its/

Comment (grahamwhaley) [reply use-cases/using-vsocks.md]: I'd also probably also swap the `,` after `implementation` to a `-` or `.`, to flow the sentence better.

File: use-cases/using-vsocks.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+|       .-------.      |
+|       | proxy |      |
+|       `-------'      |
+|           |          |
+|  .------./ \.------. |
+|  | shim |   | shim | |
+|  `------'   `------' |
+| Host                 |
+`----------------------'
+```
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: The diagram is great - v.helpful, thanks!
Sure, I might like it as a prettified .png, but it does the job as is. asciiart - so retro! ;-)

File: use-cases/using-vsocks.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+The seconds way to communicate processes and the newer one is by using
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: s/seconds/second/
s/to communicate process/for processes to communicate, and the newer one,/

File: use-cases/using-vsocks.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+|  .------.   .------. |
+|  | shim |   | shim | |
+|  `------'   `------' |
+| Host                 |
+`----------------------'
+```
+
+## System requirements
+
+Host linux kernel version must be greater than or equal to 4.8 and the
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: s/4.8/v4.8/

File: use-cases/using-vsocks.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+Kata Containers version must be greater than or equal to 1.2.0 and `use_vsock`
+must be `true` in the runtime [configuration file][1].
+
+## Advantages of using vsocks
+
+### High density
+
+Having no proxy running in the host multiplexing the connections between virtual
+machine and host processes means for around 4.5MB of RAM less per [POD][5].
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: s/means for around/saves about/

File: use-cases/using-vsocks.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+running and consuming for around 2250MB of RAM. Before making the decision not to
+use vsocks, you should ask yourself, how many containers more can run with the
+memory RAM consumed by the kata-proxy's?.
+
+### Reliability
+
+Since [kata-proxy][2] is in charge of multiplexing the connections between
+virtual machine and host processes, if it dies all connections get broken. For
+example if you have a [POD][5] with 10 containers running, if kata-proxy dies
+your 10 containers would be like aliens, they are there but it's impossible to
```
Comment (grahamwhaley) [inline use-cases/using-vsocks.md]: **aliens** - arf! :alien:

File: use-cases/using-vsocks.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+### High density
+
+Having no proxy running in the host multiplexing the connections between virtual
+machine and host processes means for around 4.5MB of RAM less per [POD][5].
+When we talk about density each kilobyte matters and it might be the decisive
+factor between run another container or not. For example if you have 500
+PODs running in a server, the same amount of [kata-proxy][2] processes will be
+running and consuming for around 2250MB of RAM. Before making the decision not to
+use vsocks, you should ask yourself, how many containers more can run with the
+memory RAM consumed by the kata-proxy's?.
```
Comment (caoruidong) [inline use-cases/using-vsocks.md]: proxies?

**Comment by devimc** [comment]
@grahamwhaley @caoruidong  changes applied, thanks

**Comment by devimc** [comment]
@klynnrif  @intelkevinputnam please take a look

File: VSocks.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    - [High density](#high-density)
+    - [Reliability](#reliability)
+
+## Introduction
+
+There are two different ways processes in the virtual machine can communicate
+with processes in the host. The first one is by using serial ports, where the
+processes in the virtual machine can read/write data from/to a serial port device
+and the processes in the host can read/write data from/to a unix socket. This
+solution has been part of QEMU from long time ago and almost all GNU/Linux
```
Comment (intelkevinputnam) [inline VSocks.md]: Lines 16 and 17  - Remove "This solution ... long time ago and almost all" and begin sentence with "Most GNU/Linux ..."

File: VSocks.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+distributions have support for serial ports, making it the most portable solution.
+However, there is a limitation that impact its usage and implementation. The
+communication between processes is serial, making impossible that two or more
```
Comment (intelkevinputnam) [inline VSocks.md]: Lines 19, 20, and 21 - "However, the serial link limits read/write access to one process at a time."

File: VSocks.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+|           |          |
+|  .------./ \.------. |
+|  | shim |   | shim | |
+|  `------'   `------' |
+| Host                 |
+`----------------------'
+```
+
+The second way for processes to communicate, and the newer one, is by using
+[vsocks][4]. The communication through vsocks is not serial and can accept
```
Comment (intelkevinputnam) [inline VSocks.md]: Lines 57, 58, and 59 - Try "A newer, simpler method is [vsocks][4], which can accept connections from multiple clients and does not require multiplexers ..."

File: VSocks.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+|  .------.   .------. |
+|  | shim |   | shim | |
+|  `------'   `------' |
+| Host                 |
+`----------------------'
+```
+
+## System requirements
+
+Host linux kernel version must be greater than or equal to v4.8 and the
```
Comment (intelkevinputnam) [inline VSocks.md]: Lines 90 and 91 - "The host Linux kernel version must be greater than or equal to v4.8, and ..."

File: VSocks.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+`vhost_vsock` module must be loaded or built-in (CONFIG_VHOST_VSOCK=y). To
+load the module run the following command:
+
+```
+$ sudo modprobe -i vhost_vsock
+```
+
+Kata Containers version must be greater than or equal to 1.2.0 and `use_vsock`
```
Comment (intelkevinputnam) [inline VSocks.md]: Line 98 - The Kata Containers ... 1.2.0, and ... be set to `true` ..."

File: VSocks.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+must be `true` in the runtime [configuration file][1].
+
+## Advantages of using vsocks
+
+### High density
+
+Having no proxy running in the host multiplexing the connections between virtual
```
Comment (intelkevinputnam) [inline VSocks.md]: Lines 105-110 - How about "Using a proxy for multiplexing the connections between the VM and the host uses 4.5MB per [POD][5]. In a high density deployment this could add up to GBs of memory that could have been used to host more PODs."

File: VSocks.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+machine and host processes saves about 4.5MB of RAM less per [POD][5].
+When we talk about density each kilobyte matters and it might be the decisive
+factor between run another container or not. For example if you have 500
+PODs running in a server, the same amount of [kata-proxy][2] processes will be
+running and consuming for around 2250MB of RAM. Before making the decision not to
+use vsocks, you should ask yourself, how many containers more can run with the
```
Comment (intelkevinputnam) [inline VSocks.md]: Line 111 - "how many more containers"

File: VSocks.md [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+memory RAM consumed by the kata-proxies?.
+
+### Reliability
+
+Since [kata-proxy][2] is in charge of multiplexing the connections between
+virtual machine and host processes, if it dies all connections get broken. For
+example if you have a [POD][5] with 10 containers running, if kata-proxy dies
+your 10 containers would be like aliens, they are there but it's impossible to
```
Comment (intelkevinputnam) [inline VSocks.md]: Line 119 - "it would be impossible to contact your containers, though they would still be running."

File: VSocks.md [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+contact them.
+As was mentioned in the [Introduction](#introduction), when vsocks are used to
```
Comment (intelkevinputnam) [inline VSocks.md]: Line 121-124 - How about "Since communication via vsocks is direct, the only way to lose communication with the virtual machine is if the VM itself or the [shim][6] dies."

**Comment by devimc** [comment]
@intelkevinputnam changes applied, thanks

**Comment by devimc** [comment]
@kata-containers/documentation  please take a look

**Comment by jodh-intel** [comment]
lgtm

**Comment by GabyCT** [comment]
lgtm
